library/core/src/str/validations/utf8_dfa.rs

```diff
@@ -0,0 +1,324 @@
+    if backup {
+        pos = backup_not_yet_invalid(input, pos);
+    }
+    validate_carefully(input, pos).unwrap_err()
+}
+
+#[inline]
+fn skip_ascii_chunks(s: &mut core::slice::Iter<'_, [u8; CHUNK_LEN]>) -> usize {
+    let mut i = 0;
+    let initial_slice = s.as_slice();
+    while let Some(c) = s.next() {
+        if !all_ascii_chunk(c) {
+            break;
+        }
+        i += 1;
+    }
+    *s = initial_slice[i..].iter();
+    i
+}
+
+#[inline]
+fn all_ascii_chunk(s: &[u8; CHUNK_LEN]) -> bool {
+    // Sadly, `core::simd` currently does not compile very efficiently on some
+    // targets (all the targets without simd, and some of the targets with it).
+    //
+    // It's also somewhat untested on others, so out of an abundance of caution
+    // we avoid it on any target that isn't both:
+    // - Known to support it efficiently.
+    // - Actually something we'd use and test on in the versions of libcore we
+    //   ship.
+    const SIMD_ASCII_TEST: bool = cfg!(any(
+        all(any(target_arch = "x86_64", target_arch = "x86"), target_feature = "sse2",),
+        all(target_arch = "aarch64", target_feature = "neon"),
+    ));
+
+    if SIMD_ASCII_TEST {
+        use crate::simd::*;
+        // Workaround for <https://github.com/rust-lang/portable-simd/issues/321> :(
+        let simd_chunk = Simd::<u8, CHUNK_LEN>::from_array(*s);
+        if cfg!(target_arch = "aarch64") {
+            simd_chunk.reduce_max() < 0x80
+        } else {
+            const ALL_HI: Simd<u8, CHUNK_LEN> = Simd::from_array([0x80; CHUNK_LEN]);
+            const ZERO: Simd<u8, CHUNK_LEN> = Simd::from_array([0; CHUNK_LEN]);
+            (simd_chunk & ALL_HI).simd_eq(ZERO).all()
+        }
+    } else {
+        // On targets where `core::simd` doesn't compile to efficient code we
+        // manually do the equivalent using u64-based SWAR using u64. Using u64 and
+        // not `usize` here seems better on 32 bit which have 64 bit register
+        // access, but ends up just being an extra unroll step on ones which don't
+        // (so no worse, and possibly still better).
+        type SwarWord = u64;
+        const WORD_BYTES: usize = core::mem::size_of::<SwarWord>();
+        const _: () = assert!((CHUNK_LEN % WORD_BYTES) == 0 && CHUNK_LEN != 0);
+        let (arr, rem) = s.as_chunks::<WORD_BYTES>();
+        debug_assert!(rem.is_empty() && !arr.is_empty());
+        let mut combined = 0;
+        for word_bytes in arr {
+            combined |= SwarWord::from_ne_bytes(*word_bytes);
+        }
+        const ALL_HI: SwarWord = SwarWord::from_ne_bytes([0x80; WORD_BYTES]);
+        (combined & ALL_HI) == 0
+    }
+}
+
+#[inline]
+fn is_ascii_small(s: &[u8]) -> bool {
+    // LLVM seems to get pretty aggressive if we use a loop here, even if we
+    // check the length first, which can end up having some pretty disasterous
+    // impacts on performance, seemingly due to inlining(?). In any case.
+    //
+    // Note that doing this for many sizes
```

Comment:
Forgot to finish this comment. Was going to mention that this seems to still beat the naïve version even if the branch on length is hard to predict, but that stops holding if you add more length conditions.